### ISSUE TITLE
Rust warns pub glob export is unused

### ISSUE THREAD

**Issue by orlp** [rust-lang/rust#160691]
I tried this code:

```rust
#[cfg(feature = "dtype-categorical")]
pub use crate::chunked_array::logical::categorical::*;
```

I expected to see this happen: no warning, since it is a pub export in a pub module.

Instead, a warning is raised:

```rust
warning: unused import: `crate::chunked_array::logical::categorical::*`
  --> crates/polars-core/src/prelude.rs:30:9
   |
30 | pub use crate::chunked_array::logical::categorical::*;
```

Reproduces in https://github.com/orlp/polars/commit/062604772a0e99738c017ca8c558d6f2d2811798, with the command

```
cargo check -p polars-core --features dtype-categorical 2>&1 | rg "pub use" --before-context=3
```

There are other warnings due to an incomplete set of features, but the specific warning here is invalid regardless of features.

### Meta

`rustc --version --verbose`:
```
rustc 1.99.0-nightly (7608eb7b0 2026-08-05)
binary: rustc
commit-hash: 7608eb7b07eaf93f16d7cf5bcb2098eca87503df
commit-date: 2026-08-05
host: aarch64-apple-darwin
release: 1.99.0-nightly
LLVM version: 23.1.0
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by orlp** [comment]
This is a regression from `nightly-2026-04-01`, currently bisecting.

**Comment by orlp** [comment]
I added `#[deny(unused_imports)` to the relevant line, giving the following bisect:

searched nightlies: from nightly-2026-04-01 to nightly-2026-08-07
regressed nightly: nightly-2026-06-09
searched commit range: https://github.com/rust-lang/rust/compare/f20a92ec01483dc5c58e90e246f266bdad822d86...cb46fbb8c6ea799c6fba9188ed889275c35a8c28
regressed commit: https://github.com/rust-lang/rust/commit/029c9e18dd1f4668e1d42bb187c1c263dfe20093

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.11</summary>


Host triple: aarch64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --start=2026-04-1 -- check -p polars-core --features dtype-categorical 
```
</details>

**Comment by orlp** [comment]
Based on that bisect, from that commit it is likely https://github.com/rust-lang/rust/pull/145108. By @LorrensP-2158466, reviewed by @petrochenkov.

I find it a bit concerning that this is the third regression in Polars in a row (https://github.com/rust-lang/rust/issues/152004, https://github.com/rust-lang/rust/issues/157420 are the previous two) when trying to bump to the latest nightly, all on import warnings. I understand that Polars is a rather extreme case of glob imports (something we'd like to move away from but haven't had the resources yet to do so), but it does indicate some sort of lack of testing in this area.

**Comment by petrochenkov** [comment]
>I find it a bit concerning that this is the third regression in Polars in a row (https://github.com/rust-lang/rust/issues/152004, https://github.com/rust-lang/rust/issues/157420 are the previous two)

The core reason for all of them is the same, polars uses many (potentially) ambiguous glob imports in creative ways + rustc has been trying to fix bugs in this area.

>but it does indicate some sort of lack of testing in this area.

Polars is our testing 😅 (And the rest of the crates.io ecosystem in general.)
All the changes were accompanied with tests of course, but with glob imports there are *a lot* of possibilities for things to go wrong.
It would be good to extend the testing with some fuzzing, but similarly, there's a problem with resources.

**Comment by apiraino** [comment]
Discussed in triage on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202026-08-13/near/616323157). Seems this issue is part of the side-effects that will kick in once we land the next stable release. #145108 was acceptd from language team and we will be adding this to the release notes (that's our way to announce breaking changes)

Unless someone has bandwidth to intervene here, I think we could close this issue.

@rustbot label -p-high

**Comment by theemathas** [comment]
@apiraino I don't think T-lang accepted that a lint should fire even when removing a reexport would change the crate's public API.

**Comment by petrochenkov** [comment]
>I expected to see this happen: no warning, since it is a pub export in a pub module.

A `pub` reexport can be reported as unused if it does not reexport anything, or reexports only something that is already reexported.

What breaks if that reexport is removed?
I tried removing it and running `cargo check --workspace --all-targets --features dtype-categorical`, and it passed.
It would be good to have some minimized reproduction.

**Comment by orlp** [comment]
> or reexports only something that is already reexported

I was not aware that's how it worked. If that's intended then this can be closed as inspecting it closer there does appear to be another path by which it is publicly exported. I apologize for the wasted effort.
